### PR TITLE
Feature/scoring output

### PR DESCRIPTION
**Pull request by Derstilon**
image
image
image
image
Added exponential number input display.
Extended amount of labels in outliner.
Added configuration of Output global settings (trace and primaries multiplier)
Added basic configuration of output quantities.

### CODE REVIEW

**Comment by grzanka** [comment]
It explodes on adding new Detect

image

**Comment by grzanka** [comment]
How to score Dose and Fluence ?

**Comment by grzanka** [comment]
List of available Quantities can be extracted from here:
https://github.com/DataMedSci/pymchelper/blob/master/pymchelper/shieldhit/detector/detector_type.py  (no 1..62)

**Comment by grzanka** [comment]
List of Detectors can be found in Table 3.9 http://shieldhit.org/download/SHIELDHIT12A_UsersGuide.pdf

**Comment by grzanka** [comment]
Detectors should appear in UI not under keyword, but under detector label

**Comment by grzanka** [comment]
Next question: can we disable Filter checkbox if no filters are being defined ?

**Comment by Derstilon** [comment]
> Next question: can we disable Filter checkbox if no filters are being defined?

done

> Detectors should appear in UI not under keyword, but under detector label

I don't understand. Can you explain it in more detail?


> How to score Dose and Fluence ?

I wasn't implemented yet but it is now.

**Comment by grzanka** [comment]
Default scored quantity should be Dose:
image

**Comment by grzanka** [comment]
Default rescale value should be 1.0:

image

**Comment by grzanka** [comment]
The could be only 2 differential scorers, 3 or more are not allowed

image

**Comment by grzanka** [comment]
Units should be added to input fields for differential scorers:
image

in a similar way as done in other places:
image

**Comment by grzanka** [comment]
Filter without any "scoring rules" are not allowed

image

**Comment by grzanka** [comment]
> > Detectors should appear in UI not under keyword, but under detector label
> 
> I don't understand. Can you explain it in more detail?

Replace "Keyword" with "Quantity" as shown below
image

File: src/ThreeEditor/util/CSG/CSGZone.ts [hunk 1]
```diff
@@ -43,61 +42,58 @@ export class Zone extends SimulationMesh {
 		CSGManagerStateChanged: Signal;
 	};
 	readonly isZone: true = true;
+	set unionOperations(operations: OperationTuple[][]) {
+		this._unionOperations = operations;
+		// If operations are specified, we have to generate fist geometry manually.
+		if (operations.length) this.updateGeometry();
```
Comment (ostatni5) [inline src/ThreeEditor/util/CSG/CSGZone.ts:48]: if we set empty operations list zone preview wont refresh

File: src/ThreeEditor/util/CSG/CSGZone.ts [hunk 2]
```diff
@@ -124,6 +120,7 @@ export class Zone extends SimulationMesh {
 	}
 
 	updateUnion(unionIndex: number, operations: OperationTuple[]): void {
+		console.log('updateUnion', unionIndex, operations);
```
Comment (ostatni5) [inline src/ThreeEditor/util/CSG/CSGZone.ts:123]:
```suggestion
```

File: src/ThreeEditor/util/CSG/CSGZone.ts [hunk 3]
```diff
@@ -108,7 +104,7 @@ export class Zone extends SimulationMesh {
 			return result.union(rowResult);
 		}, new CSG());
 
-		this.geometry.dispose();
+		this.geometry?.dispose();
```
Comment (ostatni5) [inline src/ThreeEditor/util/CSG/CSGZone.ts:107]: Geometry can be undefined?

File: src/ThreeEditor/util/Scoring/ScoringOutput.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	readonly isOutput: true = true;
+	readonly notMovable = true;
+	readonly notRotatable = true;
+	readonly notScalable = true;
+	readonly notHidable = true;
+	private _geometry?: string;
+	private _primaries: [boolean, number | null];
+	private _disabledChildren: ScoringQuantity[] = [];
+
+	//TODO: signal DetectFilter removed => clear trace[1] uuid if its the same as the removed filter
```
Comment (ostatni5) [inline src/ThreeEditor/util/Scoring/ScoringOutput.ts:33]: make issue and add number in comment

File: src/ThreeEditor/util/Scoring/ScoringQuantity.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	rescale?: number;
+};
+
+export class ScoringQuantity extends SimulationObject3D {
+	readonly isQuantity: true = true;
+	readonly notMovable = true;
+	readonly notRotatable = true;
+	readonly notScalable = true;
+	readonly notHidable = true;
+	//TODO: signal DetectFilter removed => clear _filter uuid if its the same as the removed filter
```
Comment (ostatni5) [inline src/ThreeEditor/util/Scoring/ScoringQuantity.ts:22]: add issue and mention it here

File: src/ThreeEditor/util/Ui/Number.js [hunk 6]
```diff
@@ -41,6 +59,7 @@ function createNumberInput(params) {
  *      value?: number,
  *      precision?: number
  *      update: ()=> void,
+ * 		nudge?: number,
```
Comment (ostatni5) [inline src/ThreeEditor/util/Ui/Number.js:62]: <--

**Review by ostatni5 — APPROVED** [review]
Adress comments...
obraz

**Comment by Derstilon** [comment]
> Units should be added to input fields for differential scorers: image
> 
> in a similar way as done in other places: image

image
What is the unit for this type?
I will add these at a later date because it's a minor issue and we don't have much time left this year.

**Comment by grzanka** [comment]
> image What is the unit for this type? I will add these at a later date because it's a minor issue and we don't have much time left this year.

There is no unit for Z

**Comment by grzanka** [comment]
It doesn't make sense to add Primaries checkbox if no detectors are added:
image

**Comment by grzanka** [comment]
Primaries should be integer:
image

**Comment by grzanka** [comment]
> image What is the unit for this type? I will add these at a later date because it's a minor issue and we don't have much time left this year.

Perfectly fine with me.

**Review by grzanka — APPROVED** [review]
LGTM

**Pull request merged**